### ISSUE TITLE
Custom Poisson Loss function

### ISSUE THREAD

**Issue by dzhavaharnal** [catboost/catboost#2514]
Знаю, что в catboost уже реализована Poisson loss функция. Хотел попробовать ее воспроизвести. Воспользовался учебным примером https://github.com/catboost/catboost/blob/24aceedd3abf0ddb2f4791d4b8e054deb6de916f/catboost/tutorials/custom_loss/custom_loss_and_metric_tutorial.ipynb

Получился вот такой код для CustomObjective и CustomMetric:

class CustomPoissonObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)
        
        result = []
        for index in range(len(targets)):
            e = np.exp(approxes[index])
            der1 = e - targets[index]
            der2 = e

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index]

            result.append((der1, der2))
        return result

class CustomPoissonMetric(object):
    def is_max_optimal(self):
        return False

    def evaluate(self, approxes, target, weight):
        assert len(approxes) == 1
        assert len(target) == len(approxes[0])

        approx = approxes[0]

        error_sum = 0.0
        weight_sum = 0.0
        for i in range(len(approx)):
            w = 1.0 if weight is None else weight[i]
            weight_sum += w
            error_sum += w * (np.exp(approx[i]) - target[i] * approx[i])

        return error_sum, weight_sum

    def get_final_error(self, error, weight):
        return error / (weight + 1e-38)

При попытке воспроизвести эксперимент со встроенной loss_fuction='Poisson' и c loss_fuction=CustomPoissonObjective() получил разные результаты. 

Не пойму, где я ошибаюсь. Буду очень признателен за помощь!

**Comment by dzhavaharnal** [comment]
I've found that to solve problem when hessian equals zero in other libraries (xgboost, lightGBM) is added max_delta_step: der2 = np.exp(approxes[index] +max_delta_step) where default value for max_delta_step is 0.7.
